### DISCUSSION TITLE
What is the replacement method for `ZLayer.fromService` in ZIO 2?

### DISCUSSION THREAD

**Discussion by jeffwji** [zio/zio#7009] [ZIO Users]
As the topic, I'm trying to migrate some code from ZIO 1 to 2, but can't find this method from the new ZLayer. Or maybe I have any misunderstanding with ZIO 2?

**Comment by artur-jablonski** [comment]
Perhaps this is useful: https://zio.dev/guides/migrate/zio-2.x-migration-guide#constructing-layers

**Reply by jeffwji** [reply]
Thanks!

**Comment by adamgfraser** [comment]
@jeffwji We moved to a more uniform way of constructing layers in ZIO 2.0 so now if a service depends on other services you access them using `ZIO.service` in a `for` comprehension and then you can do any other necessary setup. So this:

```scala
ZLayer.fromService[Service] { service =>
  MyService(service)
}
```

Becomes:

```scala
ZLayer {
  for {
    service <- ZIO.service[Service]
  } yield MyService(service)
}
```

This provides a more uniform way of constructing layers and scales to needing to perform ZIO workflows to acquire your service or adding finalizers and using `ZLayer.scoped`.

**Reply by jeffwji** [reply]
Thanks, I realized that.

**Reply by artur-jablonski** [reply]
> @jeffwji We moved to a more uniform way of constructing layers in ZIO 2.0 so now if a service depends on other services you access them using `ZIO.service` in a `for` comprehension and then you can do any other necessary setup. So this:
> 
> ```scala
> ZLayer.fromService[Service] { service =>
>   MyService(service)
> }
> ```
> 
> Becomes:
> 
> ```scala
> ZLayer {
>   for {
>     service <- ZIO.service[Service]
>   } yield MyService(service)
> }
> ```
> 
> This provides a more uniform way of constructing layers and scales to needing to perform ZIO workflows to acquire your service or adding finalizers and using `ZLayer.scoped`.

There's also 
```scala
ZLayer.fromFunction(service => MyService(service))
```
that could be used here (I think)

**Reply by adamgfraser** [reply]
Yes definitely. `ZLayer.fromFunction(MyService(_))` is great when your service doesn't require any acquisition or finalization logic. I usually point people to `ZLayer.apply` and `ZLayer.scoped` first because those "scale with you" a little more where you can use the same pattern as your code gets more advanced. But if you don't need to do any of that then `ZLayer.fromFunction` is the other key constructor and makes it very succinct!
